// sp-pnp-js/lib/types/locale.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Locale;
(function (Locale) {
    Locale[Locale["AfrikaansSouthAfrica"] = 1078] = "AfrikaansSouthAfrica";
    Locale[Locale["AlbanianAlbania"] = 1052] = "AlbanianAlbania";
    Locale[Locale["Alsatian"] = 1156] = "Alsatian";
    Locale[Locale["AmharicEthiopia"] = 1118] = "AmharicEthiopia";
    Locale[Locale["ArabicSaudiArabia"] = 1025] = "ArabicSaudiArabia";
    Locale[Locale["ArabicAlgeria"] = 5121] = "ArabicAlgeria";
    Locale[Locale["ArabicBahrain"] = 15361] = "ArabicBahrain";
    Locale[Locale["ArabicEgypt"] = 3073] = "ArabicEgypt";
    Locale[Locale["ArabicIraq"] = 2049] = "ArabicIraq";
    Locale[Locale["ArabicJordan"] = 11265] = "ArabicJordan";
    Locale[Locale["ArabicKuwait"] = 13313] = "ArabicKuwait";
    Locale[Locale["ArabicLebanon"] = 12289] = "ArabicLebanon";
    Locale[Locale["ArabicLibya"] = 4097] = "ArabicLibya";
    Locale[Locale["ArabicMorocco"] = 6145] = "ArabicMorocco";
    Locale[Locale["ArabicOman"] = 8193] = "ArabicOman";
    Locale[Locale["ArabicQatar"] = 16385] = "ArabicQatar";
    Locale[Locale["ArabicSyria"] = 10241] = "ArabicSyria";
    Locale[Locale["ArabicTunisia"] = 7169] = "ArabicTunisia";
    Locale[Locale["ArabicUAE"] = 14337] = "ArabicUAE";
    Locale[Locale["ArabicYemen"] = 9217] = "ArabicYemen";
    Locale[Locale["ArmenianArmenia"] = 1067] = "ArmenianArmenia";
    Locale[Locale["Assamese"] = 1101] = "Assamese";
    Locale[Locale["AzeriCyrillic"] = 2092] = "AzeriCyrillic";
    Locale[Locale["AzeriLatin"] = 1068] = "AzeriLatin";
    Locale[Locale["Bashkir"] = 1133] = "Bashkir";
    Locale[Locale["Basque"] = 1069] = "Basque";
    Locale[Locale["Belarusian"] = 1059] = "Belarusian";
    Locale[Locale["BengaliIndia"] = 1093] = "BengaliIndia";
    Locale[Locale["BengaliBangladesh"] = 2117] = "BengaliBangladesh";
    Locale[Locale["BosnianBosniaHerzegovina"] = 5146] = "BosnianBosniaHerzegovina";
    Locale[Locale["Breton"] = 1150] = "Breton";
    Locale[Locale["Bulgarian"] = 1026] = "Bulgarian";
    Locale[Locale["Burmese"] = 1109] = "Burmese";
    Locale[Locale["Catalan"] = 1027] = "Catalan";
    Locale[Locale["CherokeeUnitedStates"] = 1116] = "CherokeeUnitedStates";
    Locale[Locale["ChinesePeoplesRepublicofChina"] = 2052] = "ChinesePeoplesRepublicofChina";
    Locale[Locale["ChineseSingapore"] = 4100] = "ChineseSingapore";
    Locale[Locale["ChineseTaiwan"] = 1028] = "ChineseTaiwan";
    Locale[Locale["ChineseHongKongSAR"] = 3076] = "ChineseHongKongSAR";
    Locale[Locale["ChineseMacaoSAR"] = 5124] = "ChineseMacaoSAR";
    Locale[Locale["Corsican"] = 1155] = "Corsican";
    Locale[Locale["Croatian"] = 1050] = "Croatian";
    Locale[Locale["CroatianBosniaHerzegovina"] = 4122] = "CroatianBosniaHerzegovina";
    Locale[Locale["Czech"] = 1029] = "Czech";
    Locale[Locale["Danish"] = 1030] = "Danish";
    Locale[Locale["Dari"] = 1164] = "Dari";
    Locale[Locale["Divehi"] = 1125] = "Divehi";
    Locale[Locale["DutchNetherlands"] = 1043] = "DutchNetherlands";
    Locale[Locale["DutchBelgium"] = 2067] = "DutchBelgium";
    Locale[Locale["Edo"] = 1126] = "Edo";
    Locale[Locale["EnglishUnitedStates"] = 1033] = "EnglishUnitedStates";
    Locale[Locale["EnglishUnitedKingdom"] = 2057] = "EnglishUnitedKingdom";
    Locale[Locale["EnglishAustralia"] = 3081] = "EnglishAustralia";
    Locale[Locale["EnglishBelize"] = 10249] = "EnglishBelize";
    Locale[Locale["EnglishCanada"] = 4105] = "EnglishCanada";
    Locale[Locale["EnglishCaribbean"] = 9225] = "EnglishCaribbean";
    Locale[Locale["EnglishHongKongSAR"] = 15369] = "EnglishHongKongSAR";
    Locale[Locale["EnglishIndia"] = 16393] = "EnglishIndia";
    Locale[Locale["EnglishIndonesia"] = 14345] = "EnglishIndonesia";
    Locale[Locale["EnglishIreland"] = 6153] = "EnglishIreland";
    Locale[Locale["EnglishJamaica"] = 8201] = "EnglishJamaica";
    Locale[Locale["EnglishMalaysia"] = 17417] = "EnglishMalaysia";
    Locale[Locale["EnglishNewZealand"] = 5129] = "EnglishNewZealand";
    Locale[Locale["EnglishPhilippines"] = 13321] = "EnglishPhilippines";
    Locale[Locale["EnglishSingapore"] = 18441] = "EnglishSingapore";
    Locale[Locale["EnglishSouthAfrica"] = 7177] = "EnglishSouthAfrica";
    Locale[Locale["EnglishTrinidad"] = 11273] = "EnglishTrinidad";
    Locale[Locale["EnglishZimbabwe"] = 12297] = "EnglishZimbabwe";
    Locale[Locale["Estonian"] = 1061] = "Estonian";
    Locale[Locale["Faroese"] = 1080] = "Faroese";
    Locale[Locale["Farsi"] = 1065] = "Farsi";
    Locale[Locale["Filipino"] = 1124] = "Filipino";
    Locale[Locale["Finnish"] = 1035] = "Finnish";
    Locale[Locale["FrenchFrance"] = 1036] = "FrenchFrance";
    Locale[Locale["FrenchBelgium"] = 2060] = "FrenchBelgium";
    Locale[Locale["FrenchCameroon"] = 11276] = "FrenchCameroon";
    Locale[Locale["FrenchCanada"] = 3084] = "FrenchCanada";
    Locale[Locale["FrenchDemocraticRepofCongo"] = 9228] = "FrenchDemocraticRepofCongo";
    Locale[Locale["FrenchCotedIvoire"] = 12300] = "FrenchCotedIvoire";
    Locale[Locale["FrenchHaiti"] = 15372] = "FrenchHaiti";
    Locale[Locale["FrenchLuxembourg"] = 5132] = "FrenchLuxembourg";
    Locale[Locale["FrenchMali"] = 13324] = "FrenchMali";
    Locale[Locale["FrenchMonaco"] = 6156] = "FrenchMonaco";
    Locale[Locale["FrenchMorocco"] = 14348] = "FrenchMorocco";
    Locale[Locale["FrenchNorthAfrica"] = 58380] = "FrenchNorthAfrica";
    Locale[Locale["FrenchReunion"] = 8204] = "FrenchReunion";
    Locale[Locale["FrenchSenegal"] = 10252] = "FrenchSenegal";
    Locale[Locale["FrenchSwitzerland"] = 4108] = "FrenchSwitzerland";
    Locale[Locale["FrenchWestIndies"] = 7180] = "FrenchWestIndies";
    Locale[Locale["FrisianNetherlands"] = 1122] = "FrisianNetherlands";
    Locale[Locale["FulfuldeNigeria"] = 1127] = "FulfuldeNigeria";
    Locale[Locale["FYROMacedonian"] = 1071] = "FYROMacedonian";
    Locale[Locale["Galician"] = 1110] = "Galician";
    Locale[Locale["Georgian"] = 1079] = "Georgian";
    Locale[Locale["GermanGermany"] = 1031] = "GermanGermany";
    Locale[Locale["GermanAustria"] = 3079] = "GermanAustria";
    Locale[Locale["GermanLiechtenstein"] = 5127] = "GermanLiechtenstein";
    Locale[Locale["GermanLuxembourg"] = 4103] = "GermanLuxembourg";
    Locale[Locale["GermanSwitzerland"] = 2055] = "GermanSwitzerland";
    Locale[Locale["Greek"] = 1032] = "Greek";
    Locale[Locale["Greenlandic"] = 1135] = "Greenlandic";
    Locale[Locale["GuaraniParaguay"] = 1140] = "GuaraniParaguay";
    Locale[Locale["Gujarati"] = 1095] = "Gujarati";
    Locale[Locale["HausaNigeria"] = 1128] = "HausaNigeria";
    Locale[Locale["HawaiianUnitedStates"] = 1141] = "HawaiianUnitedStates";
    Locale[Locale["Hebrew"] = 1037] = "Hebrew";
    Locale[Locale["Hindi"] = 1081] = "Hindi";
    Locale[Locale["Hungarian"] = 1038] = "Hungarian";
    Locale[Locale["IbibioNigeria"] = 1129] = "IbibioNigeria";
    Locale[Locale["Icelandic"] = 1039] = "Icelandic";
    Locale[Locale["IgboNigeria"] = 1136] = "IgboNigeria";
    Locale[Locale["Indonesian"] = 1057] = "Indonesian";
    Locale[Locale["Inuktitut"] = 1117] = "Inuktitut";
    Locale[Locale["Irish"] = 2108] = "Irish";
    Locale[Locale["ItalianItaly"] = 1040] = "ItalianItaly";
    Locale[Locale["ItalianSwitzerland"] = 2064] = "ItalianSwitzerland";
    Locale[Locale["Japanese"] = 1041] = "Japanese";
    Locale[Locale["Kiche"] = 1158] = "Kiche";
    Locale[Locale["Kannada"] = 1099] = "Kannada";
    Locale[Locale["KanuriNigeria"] = 1137] = "KanuriNigeria";
    Locale[Locale["Kashmiri"] = 2144] = "Kashmiri";
    Locale[Locale["KashmiriArabic"] = 1120] = "KashmiriArabic";
    Locale[Locale["Kazakh"] = 1087] = "Kazakh";
    Locale[Locale["Khmer"] = 1107] = "Khmer";
    Locale[Locale["Kinyarwanda"] = 1159] = "Kinyarwanda";
    Locale[Locale["Konkani"] = 1111] = "Konkani";
    Locale[Locale["Korean"] = 1042] = "Korean";
    Locale[Locale["KyrgyzCyrillic"] = 1088] = "KyrgyzCyrillic";
    Locale[Locale["Lao"] = 1108] = "Lao";
    Locale[Locale["Latin"] = 1142] = "Latin";
    Locale[Locale["Latvian"] = 1062] = "Latvian";
    Locale[Locale["Lithuanian"] = 1063] = "Lithuanian";
    Locale[Locale["Luxembourgish"] = 1134] = "Luxembourgish";
    Locale[Locale["MalayMalaysia"] = 1086] = "MalayMalaysia";
    Locale[Locale["MalayBruneiDarussalam"] = 2110] = "MalayBruneiDarussalam";
    Locale[Locale["Malayalam"] = 1100] = "Malayalam";
    Locale[Locale["Maltese"] = 1082] = "Maltese";
    Locale[Locale["Manipuri"] = 1112] = "Manipuri";
    Locale[Locale["MaoriNewZealand"] = 1153] = "MaoriNewZealand";
    Locale[Locale["Mapudungun"] = 1146] = "Mapudungun";
    Locale[Locale["Marathi"] = 1102] = "Marathi";
    Locale[Locale["Mohawk"] = 1148] = "Mohawk";
    Locale[Locale["MongolianCyrillic"] = 1104] = "MongolianCyrillic";
    Locale[Locale["MongolianMongolian"] = 2128] = "MongolianMongolian";
    Locale[Locale["Nepali"] = 1121] = "Nepali";
    Locale[Locale["NepaliIndia"] = 2145] = "NepaliIndia";
    Locale[Locale["NorwegianBokm\u00E5l"] = 1044] = "NorwegianBokm\u00E5l";
    Locale[Locale["NorwegianNynorsk"] = 2068] = "NorwegianNynorsk";
    Locale[Locale["Occitan"] = 1154] = "Occitan";
    Locale[Locale["Oriya"] = 1096] = "Oriya";
    Locale[Locale["Oromo"] = 1138] = "Oromo";
    Locale[Locale["Papiamentu"] = 1145] = "Papiamentu";
    Locale[Locale["Pashto"] = 1123] = "Pashto";
    Locale[Locale["Polish"] = 1045] = "Polish";
    Locale[Locale["PortugueseBrazil"] = 1046] = "PortugueseBrazil";
    Locale[Locale["PortuguesePortugal"] = 2070] = "PortuguesePortugal";
    Locale[Locale["Punjabi"] = 1094] = "Punjabi";
    Locale[Locale["PunjabiPakistan"] = 2118] = "PunjabiPakistan";
    Locale[Locale["QuechaBolivia"] = 1131] = "QuechaBolivia";
    Locale[Locale["QuechaEcuador"] = 2155] = "QuechaEcuador";
    Locale[Locale["QuechaPeru"] = 3179] = "QuechaPeru";
    Locale[Locale["RhaetoRomanic"] = 1047] = "RhaetoRomanic";
    Locale[Locale["Romanian"] = 1048] = "Romanian";
    Locale[Locale["RomanianMoldava"] = 2072] = "RomanianMoldava";
    Locale[Locale["Russian"] = 1049] = "Russian";
    Locale[Locale["RussianMoldava"] = 2073] = "RussianMoldava";
    Locale[Locale["SamiLappish"] = 1083] = "SamiLappish";
    Locale[Locale["Sanskrit"] = 1103] = "Sanskrit";
    Locale[Locale["ScottishGaelic"] = 1084] = "ScottishGaelic";
    Locale[Locale["Sepedi"] = 1132] = "Sepedi";
    Locale[Locale["SerbianCyrillic"] = 3098] = "SerbianCyrillic";
    Locale[Locale["SerbianLatin"] = 2074] = "SerbianLatin";
    Locale[Locale["SindhiIndia"] = 1113] = "SindhiIndia";
    Locale[Locale["SindhiPakistan"] = 2137] = "SindhiPakistan";
    Locale[Locale["SinhaleseSriLanka"] = 1115] = "SinhaleseSriLanka";
    Locale[Locale["Slovak"] = 1051] = "Slovak";
    Locale[Locale["Slovenian"] = 1060] = "Slovenian";
    Locale[Locale["Somali"] = 1143] = "Somali";
    Locale[Locale["Sorbian"] = 1070] = "Sorbian";
    Locale[Locale["SpanishSpainModernSort"] = 3082] = "SpanishSpainModernSort";
    Locale[Locale["SpanishSpainTraditionalSort"] = 1034] = "SpanishSpainTraditionalSort";
    Locale[Locale["SpanishArgentina"] = 11274] = "SpanishArgentina";
    Locale[Locale["SpanishBolivia"] = 16394] = "SpanishBolivia";
    Locale[Locale["SpanishChile"] = 13322] = "SpanishChile";
    Locale[Locale["SpanishColombia"] = 9226] = "SpanishColombia";
    Locale[Locale["SpanishCostaRica"] = 5130] = "SpanishCostaRica";
    Locale[Locale["SpanishDominicanRepublic"] = 7178] = "SpanishDominicanRepublic";
    Locale[Locale["SpanishEcuador"] = 12298] = "SpanishEcuador";
    Locale[Locale["SpanishElSalvador"] = 17418] = "SpanishElSalvador";
    Locale[Locale["SpanishGuatemala"] = 4106] = "SpanishGuatemala";
    Locale[Locale["SpanishHonduras"] = 18442] = "SpanishHonduras";
    Locale[Locale["SpanishLatinAmerica"] = 22538] = "SpanishLatinAmerica";
    Locale[Locale["SpanishMexico"] = 2058] = "SpanishMexico";
    Locale[Locale["SpanishNicaragua"] = 19466] = "SpanishNicaragua";
    Locale[Locale["SpanishPanama"] = 6154] = "SpanishPanama";
    Locale[Locale["SpanishParaguay"] = 15370] = "SpanishParaguay";
    Locale[Locale["SpanishPeru"] = 10250] = "SpanishPeru";
    Locale[Locale["SpanishPuertoRico"] = 20490] = "SpanishPuertoRico";
    Locale[Locale["SpanishUnitedStates"] = 21514] = "SpanishUnitedStates";
    Locale[Locale["SpanishUruguay"] = 14346] = "SpanishUruguay";
    Locale[Locale["SpanishVenezuela"] = 8202] = "SpanishVenezuela";
    Locale[Locale["Sutu"] = 1072] = "Sutu";
    Locale[Locale["Swahili"] = 1089] = "Swahili";
    Locale[Locale["Swedish"] = 1053] = "Swedish";
    Locale[Locale["SwedishFinland"] = 2077] = "SwedishFinland";
    Locale[Locale["Syriac"] = 1114] = "Syriac";
    Locale[Locale["Tajik"] = 1064] = "Tajik";
    Locale[Locale["TamazightArabic"] = 1119] = "TamazightArabic";
    Locale[Locale["TamazightLatin"] = 2143] = "TamazightLatin";
    Locale[Locale["Tamil"] = 1097] = "Tamil";
    Locale[Locale["Tatar"] = 1092] = "Tatar";
    Locale[Locale["Telugu"] = 1098] = "Telugu";
    Locale[Locale["Thai"] = 1054] = "Thai";
    Locale[Locale["TibetanBhutan"] = 2129] = "TibetanBhutan";
    Locale[Locale["TibetanPeoplesRepublicofChina"] = 1105] = "TibetanPeoplesRepublicofChina";
    Locale[Locale["TigrignaEritrea"] = 2163] = "TigrignaEritrea";
    Locale[Locale["TigrignaEthiopia"] = 1139] = "TigrignaEthiopia";
    Locale[Locale["Tsonga"] = 1073] = "Tsonga";
    Locale[Locale["Tswana"] = 1074] = "Tswana";
    Locale[Locale["Turkish"] = 1055] = "Turkish";
    Locale[Locale["Turkmen"] = 1090] = "Turkmen";
    Locale[Locale["UighurChina"] = 1152] = "UighurChina";
    Locale[Locale["Ukrainian"] = 1058] = "Ukrainian";
    Locale[Locale["Urdu"] = 1056] = "Urdu";
    Locale[Locale["UrduIndia"] = 2080] = "UrduIndia";
    Locale[Locale["UzbekCyrillic"] = 2115] = "UzbekCyrillic";
    Locale[Locale["UzbekLatin"] = 1091] = "UzbekLatin";
    Locale[Locale["Venda"] = 1075] = "Venda";
    Locale[Locale["Vietnamese"] = 1066] = "Vietnamese";
    Locale[Locale["Welsh"] = 1106] = "Welsh";
    Locale[Locale["Wolof"] = 1160] = "Wolof";
    Locale[Locale["Xhosa"] = 1076] = "Xhosa";
    Locale[Locale["Yakut"] = 1157] = "Yakut";
    Locale[Locale["Yi"] = 1144] = "Yi";
    Locale[Locale["Yiddish"] = 1085] = "Yiddish";
    Locale[Locale["Yoruba"] = 1130] = "Yoruba";
    Locale[Locale["Zulu"] = 1077] = "Zulu";
    Locale[Locale["HIDHumanInterfaceDevice"] = 1279] = "HIDHumanInterfaceDevice";
})(Locale = exports.Locale || (exports.Locale = {}));
